test/unit/util/loose-equal.spec.js

```diff
@@ -0,0 +1,174 @@
+import { looseEqual } from '../../../src/shared/util'
+
+describe('looseEqual', () => {
+  it('compares booleans correctly', () => {
+    expect(looseEqual(true, true)).toBe(true)
+    expect(looseEqual(false, false)).toBe(true)
+    expect(looseEqual(true, false)).toBe(false)
+    expect(looseEqual(true, true)).toBe(true)
+    expect(looseEqual(true, 1)).toBe(false)
+    expect(looseEqual(false, 0)).toBe(false)
+  })
+
+  it('compares strings correctly', () => {
+    const text = 'Lorem ipsum'
+    const number = 1
+    const bool = true
+
+    expect(looseEqual(text, text)).toBe(true)
+    expect(looseEqual(text, text.slice(0, -1))).toBe(false)
+    expect(looseEqual(String(number), number)).toBe(true)
+    expect(looseEqual(String(bool), bool)).toBe(true)
+  })
+
+  it('compares numbers correctly', () => {
+    const number = 100
+    const decimal = 2.5
+    const multiplier = 1.0000001
+
+    expect(looseEqual(number, number)).toBe(true)
+    expect(looseEqual(number, number - 1)).toBe(false)
+    expect(looseEqual(decimal, decimal)).toBe(true)
+    expect(looseEqual(decimal, decimal * multiplier)).toBe(false)
+    expect(looseEqual(number, number * multiplier)).toBe(false)
+    expect(looseEqual(multiplier, multiplier)).toBe(true)
+  })
+
+  it('compares dates correctly', () => {
+    const date1 = new Date(2019, 1, 2, 3, 4, 5, 6)
+    const date2 = new Date(2019, 1, 2, 3, 4, 5, 6)
+    const date3 = new Date(2019, 1, 2, 3, 4, 5, 7)
+    const date4 = new Date(2219, 1, 2, 3, 4, 5, 6)
+
+    // Identical date object references
+    expect(looseEqual(date1, date1)).toBe(true)
+    // Different date references with identical values
+    expect(looseEqual(date1, date2)).toBe(true)
+    // Dates with slightly different time (ms)
+    expect(looseEqual(date1, date3)).toBe(false)
+    // Dates with different year
+    expect(looseEqual(date1, date4)).toBe(false)
+  })
+
+  it('compares files correctly', () => {
+    const date1 = new Date(2019, 1, 2, 3, 4, 5, 6)
+    const date2 = new Date(2019, 1, 2, 3, 4, 5, 7)
+    const file1 = new File([''], 'filename.txt', { type: 'text/plain', lastModified: date1 })
```

Comment:
where is File coming from?

Reply:
What do you mean with "coming from"?

Reply:
https://stackoverflow.com/a/26181292

Reply:
sorry, I meant why is this a test?

Reply:
Because the previous implementation failed here.
It treated two completely different files as same.

This test covers that.

Reply:
The main issue was to use `Object.keys()` to compare the keys across two objects.
It returns an empty array  for e.g. File objects.